src/display/MoveCameraMouseAction.cpp

```diff
@@ -0,0 +1,97 @@
+
+bool MoveCameraMouseAction::eventFilter(QObject* watched, QEvent* event) {
+    if (watched == m_watched) {
+        if (event->type() == QEvent::MouseMove) {
+            QMouseEvent* moveCameraEvent = static_cast<QMouseEvent *>(event);
+            const int x = moveCameraEvent->x();
+            const int y = moveCameraEvent->y();
+            int globalX = moveCameraEvent->globalX();
+            int globalY = moveCameraEvent->globalY();
+
+            bool resetX = false, resetY = false;
+
+            if (prevMouseX != -1 && prevMouseY != -1 && 
+                (moveCameraEvent->buttons() & (m_watched->getRotateCamera() | m_watched->getMoveCamera()))) {
+                if (skipNextMouseMoveEvent) {
+                    skipNextMouseMoveEvent = false;
+                }
+                if (moveCameraEvent->buttons() & (m_watched->getRotateCamera())) {
+                    const bool rotateThirdAxis = QApplication::keyboardModifiers().testFlag(m_watched->getRotateAroundThirdAxisModifier());
+                    m_watched->getCamera()->processRotateRequest(x - prevMouseX, y - prevMouseY, rotateThirdAxis);
+                }
+                if (moveCameraEvent->buttons() & (m_watched->getMoveCamera())) {
+                    m_watched->getCamera()->processMoveRequest(x - prevMouseX, y - prevMouseY);
+                }
+
+                m_watched->forceRerenderFrame();
+
+                const QPoint topLeft = m_watched->mapToGlobal(QPoint(0, 0));
+                const QPoint bottomRight = m_watched->mapToGlobal(QPoint(m_watched->size().width(), m_watched->size().height()));
+
+                int newX = -1;
+                int newY = -1;
+
+                if (globalX <= topLeft.x()) {
+                    newX = bottomRight.x() - 1;
+                    resetX = true;
+                }
+                if (globalX >= bottomRight.x()) {
+                    newX = topLeft.x() + 1;
+                    resetX = true;
+                }
+                if (globalY <= topLeft.y()) {
+                    newY = bottomRight.y() - 1;
+                    resetY = true;
+                }
+                if (globalY >= bottomRight.y()) {
+                    newY = topLeft.y() + 1;
+                    resetY = true;
+                }
+
+                if (resetX || resetY) {
+                    prevMouseX = resetX ? m_watched->mapFromGlobal(QPoint(newX, newY)).x() : x;
+                    prevMouseY = resetY ? m_watched->mapFromGlobal(QPoint(newX, newY)).y() : y;
+                    QCursor::setPos(resetX ? newX : globalX, resetY ? newY : globalY);
+                    skipNextMouseMoveEvent = true;
+                }
+            }
+
+            if (!resetX && !resetY) {
+                prevMouseX = x;
+                prevMouseY = y;
+            }
+        }
+        else if (event->type() == QEvent::MouseButtonPress) {
+            m_watched->getDocument()->getDisplayGrid()->setActiveDisplay(m_watched);
```

Comment:
But `setActiveDisplay(m_watched);` only present in `DisplayGrid` class.

Reply:
@drossberg  Ahh my bad earlier I removed `const` but now I see why that's the problem since the return type is const(i.e `const Document* getDocument() const;`) and it is calling `getDisplayGrid()` which is `non-const` method inside `DisplayGrid` class.

Reply:
Sorry, I mismatched the classes in my first comment: Isn't `m_watched->getDocument()->getDisplayGrid()` the same as `m_parent`?

Reply:
> Sorry, I mismatched the classes in my first comment: Isn't `m_watched->getDocument()->getDisplayGrid()` the same as `m_parent`?

`m_parent`? Btw I have done something like this
```cpp
DisplayGrid* displayGrid = m_watched->getDocument()->getDisplayGrid();
displayGrid->setActiveDisplay(m_watched);
```